server/games/game_rater.py

```diff
@@ -0,0 +1,121 @@
+import trueskill
+from trueskill import Rating
+from server.rating import RatingType
+from ..decorators import with_logger
+
+from server.games.game_results import GameOutcome
+from server.config import FFA_TEAM
+
+from typing import List, Dict, Iterable
+from server.players import Player
+
+
+class GameRatingError(Exception):
+    pass
+
+
+@with_logger
+class GameRater(object):
+    def __init__(
+        self,
+        players_by_team: Dict[int, List[Player]],
+        outcome_by_player: Dict[Player, GameOutcome],
+        rating_type=RatingType.GLOBAL
+    ):
+        self._rating_type = rating_type
+        self._outcome_by_player = outcome_by_player
+        self._players_by_team = players_by_team
+
+    def compute_rating(self) -> Dict[Player, Rating]:
+        rating_groups = self._get_rating_groups()
+        team_outcomes = [
+            self._get_team_outcome(team.keys()) for team in rating_groups
+        ]
+        ranks = self._ranks_from_team_outcomes(team_outcomes)
+
+        self._logger.debug("Rating groups: %s", rating_groups)
+        self._logger.debug("Ranks: %s", ranks)
+        return trueskill.rate(rating_groups, ranks)
+
+    def _get_rating_groups(self) -> List[Dict[Player, Rating]]:
+        """
+        Converts a dictionary mapping team ids to players to the trueskill rating_group format
+        example input: {team1: [p1, p2], team2: [p3, p4]}
+        example output: [ {p1: Rating, p2: Rating}, {p3: Rating, p4: Rating} ]
+        """
+        if FFA_TEAM in self._players_by_team:
+            if (
+                len(self._players_by_team[FFA_TEAM]) == 2
+                and len(self._players_by_team) == 1
+            ):
+                return [{
+                    player: Rating(*player.ratings[self._rating_type])
+                } for player in self._players_by_team[FFA_TEAM]]
+            else:
+                raise GameRatingError(
+                    f"Attempted to rate FFA game with other than two players: {{team: players}} = {self._players_by_team}"
+                )
+        elif len(self._players_by_team) == 2:
+            return [{
+                player: Rating(*player.ratings[self._rating_type])
+                for player in team
+            } for _, team in self._players_by_team.items()]
+        else:
+            raise GameRatingError(
+                f"Attempted to rate non-FFA game with other than two teams: {{team: players}} = {self._players_by_team}"
+            )
+
+    def _get_team_outcome(self, team: Iterable[Player]) -> GameOutcome:
+        outcomes = set(self._outcome_by_player[player] for player in team)
+        outcomes.discard(GameOutcome.UNKNOWN)
+        if len(outcomes) == 0:
```

Comment:
```suggestion
        if not outcomes:
```
I believe this is the more "pythonic" way of doing it.
https://www.python.org/dev/peps/pep-0008/ (search for `if not seq`)